##Read in mapped viral read counts for each sample, merge and format into a feature table. Input: mapped read counts for each sample, generated by DNA virome QC shell script.##
Sample1 <- read.delim('Sample1counts_bt2_final.txt', sep = '\t', header = FALSE)
Sample2 <- read.delim('Sample2counts_bt2_final.txt', sep = '\t', header = FALSE)
Sample3 <- read.delim('Sample3counts_bt2_final.txt', sep = '\t', header = FALSE)
##Repeat lines 2-4for remaining samples.##
colnames(Sample1)[3] = 'Sample1'
colnames(Sample2)[3] = 'Sample2'
colnames(Sample3)[3] = 'Sample3'
##Repeat lines 6-8for remaining samples.##
Sample1 <- Sample1[-c(2,4)]
Sample2 <- Sample2[-c(2,4)]
Sample3 <- Sample3[-c(2,4)]
##Repeat lines 10-12for remaining samples.##
merged<- merge(Sample1,Sample1, by = 'V1')
merged <- merge(merged, Sample3, by = 'V1')
reformat <- merged
row.names(reformat) <- reformat$V1
reformat <- reformat[,!colnames(reformat) %in% "V1"]
no.zeros <- reformat[rowSums(reformat) > 0,] #Output: feature table to use as input for contamination analysis.
write.table(no.zeros, "merged_counts_bt2_final.txt", sep = "\t", quote = FALSE,col.names = NA, row.names = TRUE) #Write results to file.

##Identify contaminant viral contigs. Input: feature table from line 19 and text file with sample metadata.##
library(decontam)
metadata<-read.delim("Virome_Metadata.txt", row.names = 1)
no.zeros<-as.matrix(t(no.zeros))
identical(rownames(metadata),rownames(no.zeros)) #Output of this line should be TRUE.
contam.default<-isContaminant(no.zeros, method = 'prevalence', neg =metadata$any_ctrl, threshold=0.1)
write.csv(contam.default, "decontam_default.csv")
clean.default <- t(no.zeros[,!contam.default$contaminant=="TRUE"])
clean.default <- clean.default[rowSums(clean.default) > 0,] #Output: clean feature with samples and negative controls.
write.table(clean.default,"clean_default_samples_and_controls.txt", sep="\t", row.names = TRUE, col.names = NA)
clean.default.smp <- clean.default[,!metadata$any_ctrl=="TRUE"]
clean.default.smp <- clean.default.smp[rowSums(clean.default.smp) > 0,] #Output: clean feature table with samples only.
write.table(clean.default.smp,"clean_default_samples.txt", sep="\t", row.names = TRUE, col.names = NA) #Write results to file.

##Normalize the clean read counts by RPK (reads per kilobase) in each sample to account for contig length and sequencing depth: (viral contig read count/contig length in kb)x(1000/total sample read depth). Convert to relative abundance (total sum scaling). Use RPK counts or relative abundance as inputs for downstream analyses. Sum counts and relative abundances by genus or family where indicated.##

##K-means clustering, family level.##
library(factoextra)
library(cluster)
data.fam<-read.delim("RCA_clean_relative_abundance_family.txt", header = TRUE, row.names = 1) #Read in family-level viral relative abundance.
data.fam.t<-t(data.fam)
opt.clust.fam<-fviz_nbclust(data.fam.t, kmeans, k.max = 25, method = "wss") #Find optimal number of clusters by within sum of squares.
opt.clust.fam
set.seed(100)
gap_stat.fam <- clusGap(data.fam.t, FUN = kmeans, nstart = 25, K.max = 25, B = 50) #Find optimal number of clusters by gap statistic.
fviz_gap_stat(gap_stat.fam)
sil.kmeans.fam <- fviz_nbclust(data.fam.t, kmeans, k.max = 25, method = "silhouette") #Find optimal number of clusters by silhouette score.
sil.kmeans.fam
set.seed(100)
kmean.fam3<-kmeans(data.fam.t, 3, iter.max = 10, nstart = 25)
write.table(kmean.fam3$cluster, "kmeans_family_3.txt", sep = "\t") #Write cluster assignments to file.

##Calculate Shannon index, contig level.##
contig.data<-read.delim("RCA_clean_samples_RPK.txt", row.names = 1) #Input: RPK counts of viral contigs in samples (described in line 36).
contigs.t<-t(contig.data)
shannon.contigs<-diversity(contigs.t, index = 'shannon') #Output: viral Shannon index for each sample.
write.table(shannon.contigs,"RCA_shannon_contigs.txt", sep = '\t') #Write results to file.

#Read in virome metadata for samples, including cluster assignments and alpha diversity metrics as columns.##
metadata2 <- read.delim("Virome_Metadata.txt") 

#Multinomial logistic regression for associations with virome cluster.##
library(mclogit)
metadata2$Clust_figure_order<-as.factor(metadata2$Clust_figure_order)
sub<-subset(metadata2,Clust_figure_order =='Cluster_2'|Clust_figure_order =='Cluster_3')
#By sample type.##
set.seed(100)
mbl_type <- mblogit(Clust_figure_order~Sample_type,  data = metadata2)
set.seed(100)
mbl_type_sub <- mblogit(Clust_figure_order~Sample_type,  data = sub)
type_c1ref <- as.data.frame(summary_type$coefficients)
type_c1ref$reference <- c("Cluster_1")
type_c1ref$comparison <- rownames(type_c1ref)
type_c2ref <- as.data.frame(summary_type_sub$coefficients)
type_c2ref$reference <- c("Cluster_2")
type_c2ref$comparison <- rownames(type_c2ref)
statsdata_type <- rbind(type_c1ref,type_c2ref)
statsdata_type <- statsdata_type[!grepl("Intercept",statsdata_type$comparison),]
statsdata_type$p.adjusted <- p.adjust(statsdata_type$`Pr(>|z|)`, method = "BH")
write.table(statsdata_type, "mblogit_virome_type_pvals_adjusted.txt", row.names = FALSE, sep = "\t")

#Plot contig-level alpha diversity.#
richness.plot <- ggplot(metadata2, aes(x = Days_since_visit1, y = Richness)) +
  geom_point() +
  stat_smooth(method="lm", se=TRUE, span=0.5, level=0.95) +
  scale_x_continuous(breaks =c(0,180,360,540,720,750)) +
  theme_bw()
richness.plot

shannon.plot <- ggplot(metadata2, aes(x = Days_since_visit1, y = Shannon)) +
  geom_point() +
  stat_smooth(method="lm", se=TRUE, span=0.5, level=0.95) +
  scale_x_continuous(breaks =c(0,180,360,540,720,750)) +
  theme_bw()
shannon.plot

##Check for association between virome alpha diversity and sample type.##
kruskal.test(Richness ~ Sample_type, data = metadata2)
pairwise.wilcox.test(metadata2$Richness, metadata2$Sample_type, p.adjust.method = "BH")
kruskal.test(Shannon ~ Sample_type, data = metadata2)
pairwise.wilcox.test(metadata2$Shannon, metadata2$Sample_type, p.adjust.method = "BH")

#Shannon index over time with interaction term, contig level.##
shannon.interaction <- lme(Shannon ~ Days_since_visit1 + Shedder_or_control + (Days_since_visit1*Shedder_or_control), random =~1|Patient_factor, data = metadata2)
summary(shannon.interaction)

#Shannon index over time no interaction term, contig level. Use this since the interaction term is not significant.#
shannon.no.interaction <- lme(Shannon ~ Days_since_visit1 + Shedder_or_control, random =~1|Patient_factor, data = metadata2)
summary(shannon.no.interaction)
shannon_sum <- summary(shannon.no.interaction)
shannon_sum$tTable

#Richness over time with interaction term, contig level.##
richness.interaction <- lme(Richness ~ Days_since_visit1 + Shedder_or_control + (Days_since_visit1*Shedder_or_control), random =~1|Patient_factor, data = metadata2)
summary(richness.interaction)

#Richness over time no interaction term, contig level.#
richness.no.interaction <- lme(Richness ~ Days_since_visit1 + Shedder_or_control, random =~1|Patient_factor, data = metadata2)
summary(richness.no.interaction)
richness_sum <- summary(richness.no.interaction)
richness_sum$tTable

##Virome beta diversity analyses.##
library(vegan)
library(dplyr)
library(reshape2)
library(phyloseq)
library(ggplot2)
library(coin)

##Bray-Curtis dissimilarity (weighted).##
data<-read.delim("RCA_clean_samples_relative_abundance.txt", row.names = 1)
data.transposed<-t(data)
dis.bc <- vegdist(data.transposed, method = "bray")
dis.bc2<-as.matrix(dis.bc)
write.table(dis.bc2,"RCA_contigs_BC.txt", sep = '\t')
dis.bc2[lower.tri(dis.bc2, diag = TRUE)] <- NA
dis.bc2.long<-melt(dis.bc2)
write.table(dis.bc2.long, "RCA_contigs_BC_long_format.txt", sep="\t", row.names = FALSE) #Write deduplicated Bray-Curtis dissimilarity values to file.

##Sorensen dissimilarity (unweighted).##
data.binary <- data.transposed
data.binary[data.binary > 0] <- 1
dis.sor <- vegdist(data.binary, method = "bray")
dis.sor2 <-as.matrix(dis.sor)
write.table(dis.sor2,"RCA_contigs_sorensen.txt", sep = '\t')
dis.sor2[lower.tri(dis.sor2, diag = TRUE)] <- NA
dis.sor2.long<-melt(dis.sor2)
write.table(dis.sor2.long, "RCA_contigs_sorensen_long_format.txt", sep="\t", row.names = FALSE) #Write deduplicated Sorensen dissimilarity values to file.

##Permutation tests of Bray-Curtis dissimilarity within and between groups.##
stats.data <- read.delim("RCA_contigs_BC_long_format_metadata.txt", header = TRUE, sep = "\t")
stats.data$comparison_all <- as.factor(stats.data$comparison_all)
stats.data$comparison_DSC_vs_notDSC <- as.factor(stats.data$comparison_DSC_vs_notDSC)
stats.data$block <- c("block1")
stats.data$block <- as.factor(stats.data$block)
w_b.ctrls<-subset(stats.data,comparison_all =='Control.Control.within'|comparison_all =='Control.Control.between')
w_b.shed<-subset(stats.data,comparison_all =='Shedder.Shedder.within'|comparison_all =='Shedder.Shedder.between')
w.ctrl_shed<-subset(stats.data,comparison_all =='Control.Control.within'|comparison_all =='Shedder.Shedder.within')
b.ctrl_shed<-subset(stats.data,comparison_all =='Control.Control.between'|comparison_all =='Shedder.Shedder.between')
set.seed(100)
perm.w_b.ctrls<-independence_test(bray_curtis~comparison_all|block, data = w_b.ctrls, alternative = c("two.sided"), distribution=approximate(nresample=100000))
perm.w_b.shed<-independence_test(bray_curtis~comparison_all|block, data = w_b.shed, alternative = c("two.sided"), distribution=approximate(nresample=100000))
perm.w.ctrl_shed<-independence_test(bray_curtis~comparison_all|block, data = w.ctrl_shed, alternative = c("two.sided"), distribution=approximate(nresample=100000))
perm.b.ctrl_shed<-independence_test(bray_curtis~comparison_all|block, data = b.ctrl_shed, alternative = c("two.sided"), distribution=approximate(nresample=100000))
bc.test<-c("perm.w_b.ctrls","perm.w_b.shed","perm.w.ctrl_shed","perm.b.ctrl_shed")
p.val<-c(pvalue(perm.w_b.ctrls),pvalue(perm.w_b.shed),pvalue(perm.w.ctrl_shed),pvalue(perm.b.ctrl_shed))
bc.pvals<-data.frame(bc.test,p.val)
#Benjamini-Hochberg correction for multiple comparisons.#
pval.adjust <- p.adjust(bc.pvals$p.val, method = "BH")
bc.adjusted<-cbind(bc.pvals,pval.adjust)
write.table(bc.adjusted, "RCA_BC_permutation_adjusted.txt",row.names = FALSE,sep = "\t")

##Permutation tests of Sorensen dissimilarity within and between groups.##
stats.data.sor <- read.delim("RCA_contigs_sorensen_long_format_metadata.txt", header = TRUE, sep = "\t")
stats.data.sor$comparison_all <- as.factor(stats.data.sor$comparison_all)
stats.data.sor$comparison_DSC_vs_notDSC <- as.factor(stats.data.sor$comparison_DSC_vs_notDSC)
stats.data.sor$block <- c("block1")
stats.data.sor$block <- as.factor(stats.data.sor$block)
w_b.ctrls.sor<-subset(stats.data.sor,comparison_all =='Control.Control.within'|comparison_all =='Control.Control.between')
w_b.shed.sor<-subset(stats.data.sor,comparison_all =='Shedder.Shedder.within'|comparison_all =='Shedder.Shedder.between')
w.ctrl_shed.sor<-subset(stats.data.sor,comparison_all =='Control.Control.within'|comparison_all =='Shedder.Shedder.within')
b.ctrl_shed.sor<-subset(stats.data.sor,comparison_all =='Control.Control.between'|comparison_all =='Shedder.Shedder.between')
set.seed(100)
perm.w_b.ctrls.sor<-independence_test(sorensen~comparison_all|block, data = w_b.ctrls.sor, alternative = c("two.sided"), distribution=approximate(nresample=100000))
perm.w_b.shed.sor<-independence_test(sorensen~comparison_all|block, data = w_b.shed.sor, alternative = c("two.sided"), distribution=approximate(nresample=100000))
perm.w.ctrl_shed.sor<-independence_test(sorensen~comparison_all|block, data = w.ctrl_shed.sor, alternative = c("two.sided"), distribution=approximate(nresample=100000))
perm.b.ctrl_shed.sor<-independence_test(sorensen~comparison_all|block, data = b.ctrl_shed.sor, alternative = c("two.sided"), distribution=approximate(nresample=100000))
sor.test<-c("perm.w_b.ctrls.sor","perm.w_b.shed.sor","perm.w.ctrl_shed.sor","perm.b.ctrl_shed.sor")
p.val.sor<-c(pvalue(perm.w_b.ctrls.sor),pvalue(perm.w_b.shed.sor),pvalue(perm.w.ctrl_shed.sor),pvalue(perm.b.ctrl_shed.sor))
sor.pvals<-data.frame(sor.test,p.val.sor)
#Benjamini-Hochberg correction for multiple comparisons.#
pval.adjust.sor <- p.adjust(sor.pvals$p.val.sor, method = "BH")
sor.adjusted<-cbind(sor.pvals,pval.adjust.sor)
write.table(sor.adjusted, "RCA_sorensen_permutation_adjusted.txt",row.names = FALSE,sep = "\t")

##PCoA, weighted, all samples.##
set.seed(100)
feature_table<-read.delim("RCA_clean_default_samples_relative_abundance.txt", row.names = 1)
feature_table <- as.data.frame(feature_table)
dim(feature_table)
OTU=otu_table(feature_table,taxa_are_rows =TRUE)
sample_names(OTU)
metadata<-import_qiime_sample_data("Virome_Metadata.txt")
sample_names(metadata)
identical(sample_names(OTU),sample_names(metadata))
physeq =             phyloseq(OTU,metadata)
ord <- ordinate(physeq,
                method = "PCoA",
                distance = "bray")
PCoA.BC  <- plot_ordination(physeq = physeq,
                            ordination = ord,
                            shape = "Shedder_or_control", # metadata variable
                            color = "Days_since_visit1", # metadata variable
                            axes = c(1,2),
                            title='Bray Curtis PCoA') +
  theme_bw() +
  theme(text=element_text(size=20))+
  geom_point(size = 8)+
  scale_color_gradientn(colors = c("#fcc5c0", "#fa9fb5", "#f768a1", "#dd3497", "#ae017e", "#7a0177", "#49006a"), breaks = c(0,180,360,540,720,750))
PCoA.BC

##PCoA, unweighted, all samples##
set.seed(100)
feature_table[feature_table > 0] <- 1
OTU.sor=otu_table(feature_table,taxa_are_rows =TRUE)
sample_names(OTU.sor)
sample_names(metadata)
identical(sample_names(OTU.sor),sample_names(metadata))
physeq.sor =             phyloseq(OTU.sor,metadata)
ord.sor <- ordinate(physeq.sor,
                    method = "PCoA",
                    distance = "bray")
PCoA.sor <- plot_ordination(physeq = physeq.sor,
                            ordination = ord.sor,
                            shape = "Shedder_or_control", # metadata variable
                            color = "Days_since_visit1", # metadata variable
                            axes = c(1,2),
                            title='Sorensen PCoA') +
  theme_bw() +
  theme(text=element_text(size=20))+
  geom_point(size = 8)+
  scale_color_gradientn(colors = c("#fcc5c0", "#fa9fb5", "#f768a1", "#dd3497", "#ae017e", "#7a0177", "#49006a"), breaks = c(0,180,360,540,720,750))
PCoA.sor

##PCoA, weighted, subset by shedder/control status.##
physeq_ctrls <- subset_samples(physeq, Shedder_or_control=="Control")
set.seed(100)
ord_ctrls <- ordinate(physeq_ctrls,
                      method = "PCoA",
                      distance = "bray")
PCoA.BC.ctrl  <- plot_ordination(physeq = physeq_ctrls,
                                 ordination = ord_ctrls,
                                 #shape = "Sample_status_PCoA", # metadata variable
                                 color = "Days_since_visit1", # metadata variable
                                 axes = c(1,2),
                                 title='Bray Curtis PCoA') +
  theme_bw() +
  theme(text=element_text(size=20))+
  geom_point(size = 8)+
  scale_color_gradientn(colors = c("#fcc5c0", "#fa9fb5", "#f768a1", "#dd3497", "#ae017e", "#7a0177", "#49006a"), breaks = c(0,180,360,540,720,750))
PCoA.BC.ctrl

physeq_shed <- subset_samples(physeq, Shedder_or_control=="Shedder")
set.seed(100)
ord_shed <- ordinate(physeq_shed,
                     method = "PCoA",
                     distance = "bray")
PCoA.BC.shed  <- plot_ordination(physeq = physeq_shed,
                                 ordination = ord_shed,
                                 shape = "DSC_all", # metadata variable
                                 color = "Days_since_visit1", # metadata variable
                                 axes = c(1,2),
                                 title='Bray Curtis PCoA') +
  theme_bw() +
  theme(text=element_text(size=20))+
  geom_point(size = 8)+
  scale_color_gradientn(colors = c("#fcc5c0", "#fa9fb5", "#f768a1", "#dd3497", "#ae017e", "#7a0177", "#49006a"), breaks = c(0,180,360,540,720,750))+
  scale_shape_manual(values = c(17,17))
PCoA.BC.shed

##PCoA, unweighted, subset to controls or shedders##
set.seed(100)
physeq_ctrl.sor <- subset_samples(physeq.sor, Shedder_or_control=="Control")
ord_ctrl.sor <- ordinate(physeq_ctrl.sor,
                         method = "PCoA",
                         distance = "bray")
PCoA.sor.ctrl <- plot_ordination(physeq = physeq_ctrl.sor,
                                 ordination = ord_ctrl.sor,
                                 #shape = "DSC_all", # metadata variable
                                 color = "Days_since_visit1", # metadata variable
                                 axes = c(1,2),
                                 title='Sorensen') +
  theme_bw() +
  theme(text=element_text(size=20))+
  geom_point(size = 8)+
  scale_color_viridis_c(direction = -1)
PCoA.sor.ctrl

physeq_shed.sor <- subset_samples(physeq.sor, Shedder_or_control=="Shedder")
set.seed(100)
ord_shed.sor <- ordinate(physeq_shed.sor,
                         method = "PCoA",
                         distance = "bray")
PCoA.sor.shed <- plot_ordination(physeq = physeq_shed.sor,
                                 ordination = ord_shed.sor,
                                 shape = "DSC_all", # metadata variable
                                 color = "Days_since_visit1", # metadata variable
                                 axes = c(1,2),
                                 title='Sorensen') +
  theme_bw() +
  theme(text=element_text(size=20))+
  geom_point(size = 8)+
  scale_color_viridis_c(direction = -1)
PCoA.sor.shed

##PERMANOVA with Bray-Curtis dissimilarity (weighted)##
adonis.data<-read.delim("RCA_clean_samples_relative_abundance.txt", row.names = 1)
adonis.dataT<-t(adonis.data)
dis.bc <- vegdist(adonis.dataT, method = "bray")
dis.bc2<-as.matrix(dis.bc)
adonis.metadata <- read.delim("Virome_Metadata.txt", header = TRUE, sep = '\t')
identical(adonis.metadata$RCA_ID, rownames(dis.bc2))

##All samples, sample type.##
set.seed(100)
adonis_type <- adonis2(dis.bc2 ~ Sample_type, data = adonis.metadata, permutations = 999, by = "margin")
adonis_type

##All samples by ART duration and shedder/non-shedder status.##
set.seed(100)
perm <- how(nperm = 999, within = Within(type = "free"), plots = Plots(strata = adonis.metadata$Patient_factor))
adonis.test1 <- adonis2(dis.bc2 ~ Days_since_visit1*Shedder_or_control + Days_since_visit1 + Shedder_or_control, data = adonis.metadata, permutations = perm, by = "margin") #Check for interaction between variables.
adonis.test1
write.table(adonis.test1, "RCA_contigs_BC_adonis2_allSamples_interaction.txt", sep = "\t")
adonis.test2 <- adonis2(dis.bc2 ~ Days_since_visit1 + Shedder_or_control, data = adonis.metadata, permutations = perm, by = "margin")
adonis.test2
write.table(adonis.test2, "RCA_contigs_BC_adonis2_allSamples_noInteraction.txt", sep = "\t")

##PERMANOVA with Sorensen dissimilarity (unweighted).##
adonis.dataT.binary <- adonis.dataT
adonis.dataT.binary[adonis.dataT.binary > 0] <- 1
dis.sor <- vegdist(adonis.dataT.binary, method = "bray")
dis.sor<-as.matrix(dis.sor)

##All samples, sample type.##
set.seed(100)
adonis_type_sor <- adonis2(dis.sor ~ Sample_type, data = adonis.metadata, permutations = 999, by = "margin")
adonis_type_sor

set.seed(100)
adonis.sor.int <- adonis2(dis.sor ~ Days_since_visit1*Shedder_or_control + Days_since_visit1 + Shedder_or_control, data = adonis.metadata, permutations = perm, by = "margin") #Check for interaction between variables.
adonis.sor.int
write.table(adonis.sor.int, "RCA_contigs_sor_adonis2_allSamples_interaction.txt", sep = "\t")
adonis.sor <- adonis2(dis.sor ~ Days_since_visit1 + Shedder_or_control, data = adonis.metadata, permutations = perm, by = "margin")
adonis.sor
write.table(adonis.sor, "RCA_contigs_sor_adonis2_allSamples_noInteraction.txt", sep = "\t")

#Subset non-shedders.#
metadata.ctrls <- adonis.metadata[adonis.metadata$Shedder_or_control=="Control",]
BC.ctrls <- dis.bc2[,colnames(dis.bc2) %in% metadata.ctrls$RCA_ID]
BC.ctrls <- BC.ctrls[rownames(BC.ctrls) %in% metadata.ctrls$RCA_ID,]
dis.bc.ctrls <- as.dist(BC.ctrls)

#PERMANOVA of non-shedders only (Bray-Curtis dissimilarity).#
set.seed(100)
perm.ctrls <- how(nperm = 999, within = Within(type = "free"), plots = Plots(strata = metadata.ctrls$Patient_factor))
adonis.ctrls <- adonis2(dis.bc.ctrls ~ Days_since_visit1, data = metadata.ctrls, permutations = perm.ctrls, by = "margin")
adonis.ctrls
write.table(adonis.ctrls, "RCA_contigs_BC_adonis2_controls.txt",sep = "\t")

#Subset shedders.#
metadata.shed <- adonis.metadata[adonis.metadata$Shedder_or_control=="Shedder",]
BC.shed <- dis.bc2[,colnames(dis.bc2) %in% metadata.shed$RCA_ID]
BC.shed <- BC.shed[rownames(BC.shed) %in% metadata.shed$RCA_ID,]
dis.bc.shed <- as.dist(BC.shed)

#PERMANOVA of shedders only (Bray-Curtis dissimilarity).#
set.seed(100)
perm.shed <- how(nperm = 999, within = Within(type = "free"), plots = Plots(strata = metadata.shed$Patient_factor))
adonis.shed <- adonis2(dis.bc.shed ~ Days_since_visit1, data = metadata.shed, permutations = perm.shed, by = "margin") #By ART duration.
adonis.shed
write.table(adonis.shed, "RCA_contigs_BC_adonis2_shedders.txt",sep = "\t")

set.seed(100)
adonis.DSC <- adonis2(dis.bc.shed ~ DSC_all, data = metadata.shed, permutations = perm.shed, by = "margin") #By discordant shedding timepoint.
adonis.DSC
write.table(adonis.DSC, "RCA_contigs_BC_adonis2_shedders_DSC.txt",sep = "\t")

##Differential abundance analysis with MaAsLin2.##
library(Maaslin2)

#Family level analysis.#
family_table<-read.delim("RCA_clean_relative_abundance_family.txt", row.names = 1) #Input: Virus family relative abundance.
family_table<-as.data.frame(t(family_table))
metadata<-read.delim("Virome_Metadata.txt", row.names = 1)
identical(rownames(metadata), rownames(family_table)) #Output should be TRUE.
##Make variable to test for interaction.##
metadata$shedder_time = (metadata$Shedder_or_control == "Shedder") * metadata$Days_since_visit1
fit_data_family_interaction= Maaslin2(
  input_data = family_table, 
  input_metadata = metadata,
  min_abundance = 0,
  min_prevalence = 0.1, #Only analyze viruses found in at least 10% of samples.
  output = "maaslin2_family_interaction",
  transform = "none", #No transformation.
  normalization = "none", #Input data is already normalized as relative abundance (TSS).
  fixed_effects = c("Days_since_visit1", "Shedder_or_control","shedder_time"),
  random_effects = c("Patient_factor"))
##Read in results, subset interaction variable results, and recalculate q-values.##
interaction_all_results <- read.delim("maaslin2_family_interaction/all_results.tsv")
interaction_subset <- subset(interaction_all_results,metadata=="shedder_time")
interaction_subset$qval.readjust <- p.adjust(interaction_subset$pval, method = "BH")
write.table(interaction_subset,"maaslin2_family_interaction/interaction_subset_results.tsv",sep = "\t",col.names = NA)

fit_data_full_model= Maaslin2(
  input_data = family_table, 
  input_metadata = metadata,
  min_abundance = 0,
  min_prevalence = 0.1, #Only analyze viruses found in at least 10% of samples.
  output = "maaslin2/family/maaslin2_full_model",
  transform = "none", #No transformation.
  normalization = "none", #Input data is already normalized as relative abundance (TSS).
  fixed_effects = c("Days_since_visit1","Shedder_or_control"),
  random_effects = c("Patient_factor"))

##Subset variables to readjust q values.##
full_model_results <- read.delim("maaslin2/family/maaslin2_full_model/all_results.tsv")
full_model_time <- subset(full_model_results,metadata=="Days_since_visit1")
full_model_time$qval.readjust <- p.adjust(full_model_time$pval, method = "BH")
write.table(full_model_time,"maaslin2/family/maaslin2_full_model/time_all_results_readjusted.tsv",sep = "\t",col.names = NA)
full_model_shed_ctrl <- subset(full_model_results,metadata=="Shedder_or_control")
full_model_shed_ctrl$qval.readjust <- p.adjust(full_model_shed_ctrl$pval, method = "BH")
write.table(full_model_shed_ctrl,"maaslin2/family/maaslin2_full_model/shed_ctrl_all_results_readjusted.tsv",sep = "\t",col.names = NA)

##Subset shedders and non-shedders to analyze separately.##
family_table$Shedder_or_control<-metadata$Shedder_or_control
shedders <- subset(family_table,Shedder_or_control=="Shedder")
shedders <- shedders[,!names(shedders) %in% c('Shedder_or_control')]
controls <- subset(family_table,Shedder_or_control=="Control")
controls <- controls[,!names(controls) %in% c('Shedder_or_control')]

##Non-shedders over time##
fit_data_time_controls= Maaslin2(
  input_data = controls, 
  input_metadata = metadata,
  min_abundance = 0,
  min_prevalence = 0.1, #Only analyze viruses found in at least 10% of samples.
  output = "maaslin2_family_time_controls",
  transform = "none", #No transformation.
  normalization = "none", #Input data is already normalized as relative abundance (TSS).
  fixed_effects = c("Days_since_visit1"),
  random_effects = c("Patient_factor"))

##Shedders over time##
fit_data_time_shedders= Maaslin2(
  input_data = shedders, 
  input_metadata = metadata,
  min_abundance = 0,
  min_prevalence = 0.1, #Only analyze viruses found in at least 10% of samples.
  output = "maaslin2_family_time_shedders",
  transform = "none", #No transformation.
  normalization = "none", #Input data is already normalized as relative abundance (TSS).
  fixed_effects = c("Days_since_visit1"),
  random_effects = c("Patient_factor"))

##Genus level analysis.##
genus_table<-read.delim("RCA_clean_relAbund_genus.txt", row.names = 1)
genus_table<-as.data.frame(t(genus_table))
metadata<-read.delim("Virome_Metadata.txt", row.names = 1)

##Make variable to test for interaction.##
metadata$shedder_time = (metadata$Shedder_or_control == "Shedder") * metadata$Days_since_visit1
fit_data_genus_shedInteract= Maaslin2(
  input_data = genus_table, 
  input_metadata = metadata,
  min_abundance = 0,
  min_prevalence = 0.1, #Only analyze viruses found in at least 10% of samples.
  output = "maaslin2/genus/maaslin2_genus_shedInteract",
  transform = "none", #No transformation.
  normalization = "none", #Input data is already normalized as relative abundance (TSS).
  fixed_effects = c("Days_since_visit1", "Shedder_or_control","shedder_time"),
  random_effects = c("Patient_factor"))

##Read in results, subset interaction variable results, and recalculate q-values.##
shedInteract_all_results <- read.delim("maaslin2/genus/maaslin2_genus_shedInteract/all_results.tsv")
shedInteract_subset <- subset(shedInteract_all_results,metadata=="shedder_time")
shedInteract_subset$qval.readjust <- p.adjust(shedInteract_subset$pval, method = "BH")
write.table(shedInteract_subset,"maaslin2/genus/maaslin2_genus_shedInteract/shedInteract_subset_results.tsv",sep = "\t",col.names = NA)

##Subset shedders and non-shedders.##
genus_table$Shedder_or_control<-metadata$Shedder_or_control
shedders <- subset(genus_table,Shedder_or_control=="Shedder")
shedders <- shedders[,!names(shedders) %in% c('Shedder_or_control')]
controls <- subset(genus_table,Shedder_or_control=="Control")
controls <- controls[,!names(controls) %in% c('Shedder_or_control')]

##Non-shedders over time##
fit_data_time_controls= Maaslin2(
  input_data = controls, 
  input_metadata = metadata,
  min_abundance = 0,
  min_prevalence = 0.1, #Only analyze viruses found in at least 10% of samples.
  output = "maaslin2/genus/maaslin2_genus_time_controls",
  transform = "none", #No transformation.
  normalization = "none", #Input data is already normalized as relative abundance (TSS).
  fixed_effects = c("Days_since_visit1"),
  random_effects = c("Patient_factor"))

##Shedders over time##
fit_data_time_shedders= Maaslin2(
  input_data = shedders, 
  input_metadata = metadata,
  min_abundance = 0,
  min_prevalence = 0.1, #Only analyze viruses found in at least 10% of samples.
  output = "maaslin2/genus/maaslin2_genus_time_shedders",
  transform = "none", #No transformation.
  normalization = "none", #Input data is already normalized as relative abundance (TSS).
  fixed_effects = c("Days_since_visit1"),
  random_effects = c("Patient_factor"))



